smart_open/s3.py

```diff
@@ -420,30 +427,68 @@ def read(self, size=-1):
         return binary
 
 
+def _initialize_boto3(rw, session, resource, resource_kwargs):
+    """Created the required objects for accessing S3.  Ideally, they have
+    been already created for us and we can just reuse them.
+
+    We only really need one thing: the resource.  There are multiple ways of
+    getting one, in order of effort:
+
+    1) Directly from the user
+    2) From the session directly specified by the user
+    3) From an entirely new session
+
+    Once we have the resource, we no longer need the session.
+    """
+    if resource_kwargs is None:
+        resource_kwargs = {}
+
+    if resource:
+        if session:
+            logger.warning('ignoring session because resource was passed explicitly')
+        if resource_kwargs:
+            logger.warning('ignoring resource_kwargs because resource was passed explicitly')
+        rw._session = None
+        rw._resource = resource
+    elif session:
+        rw._session = session
+        rw._resource = rw._session.resource('s3', **resource_kwargs)
+        rw._resource_kwargs = resource_kwargs
+    else:
+        rw._session = boto3.Session()
+        rw._resource = rw._session.resource('s3', **resource_kwargs)
```

Comment:
How will such "caching" work in the presence of multi-threading and multi-processing?

Sounds risky, unless boto3 is safe / clever enough.

Reply:
It's not safe - boto3 explicitly mentions that resource objects aren't threadsafe/process-safe.

We could do the same thing in our docs, e.g.

> If you re-use boto3 session or resource objects with smart_open, take care not to share them across multiple threads or subprocesses, because that is unsafe (see boto3 docs).